File: 2021/b.hs
#!/usr/bin/env runghc

-- ACM-ICPC 2021 Domestic Qualifier
-- Problem B: Hundred-Cell Calculation Puzzles

import Control.Monad (when, replicateM)
import Data.Tuple (swap)
import qualified Data.Set as Set

main :: IO ()
main = do
    [w, h] <- map read . words <$> getLine :: IO [Int]
    when ((w, h) /= (0, 0)) $ do
        let k = w + h - 1
        xyn <- replicateM k (map read . words <$> getLine) :: IO [[Int]]
        putStrLn $ if solve w h xyn then "YES" else "NO"
        main

solve :: Int -> Int -> [[Int]] -> Bool
solve w h xyn = dfs Set.empty [1]
    where xy = [(x, y + w) | [x, y, _] <- xyn]
          adj = xy ++ map swap xy  -- bidirectional adjacency list
          dfs visited [] = Set.size visited == w + h  -- has visited all?
          dfs visited (x : stack) =
              let ys = filter (not.(`Set.member` visited)) $ lookupAll x adj
              in  dfs (Set.insert x visited) (ys ++ stack)

lookupAll :: Eq k => k -> [(k,v)] -> [v]
lookupAll k = map snd . filter ((== k) . fst)

File: 2021/b.ex
#!/usr/bin/env elixir

# ACM-ICPC 2021 Domestic Qualifier
# Problem B - Hundred-Cell Calculation Puzzles

defmodule Main do
  def split_into_list(str) do
    String.split(str, ~r{\s}, trim: true) |> Enum.map(&String.to_integer/1)
  end

  def main do
    [w, h] = IO.gets("") |> split_into_list
    if [w, h] != [0, 0] do
      k = w + h - 1
      xyn = for _ <- 1..k, do: IO.gets("") |> split_into_list
      (if solve(w, h, xyn), do: "YES", else: "NO") |> IO.puts
      main()
    end
  end

  def solve(w, h, xyn) do
    # 上端の数を頂点1, 2, ..., w, 左端の数を頂点-1, -2, ..., -hとし,
    # 解答マス(x, y)に数が入っているとき頂点x, -y間に辺があるグラフを
    # 考える. このグラフが連結であるとき答は真.
    xy  = for [x, y, _] <- xyn, do: {x, -y}  # edges between x and -y
    adj = xy ++ Enum.map(xy, fn {a, b} -> {b, a} end)  # bidirectional
    # 頂点1から全頂点に到達可能か?
    w + h == MapSet.size(dfs(MapSet.new(), adj, 1))
  end

  # startから深さ優先探索
  def dfs(visited, adj, start) do
    neighbors = for {a, b} <- adj, a == start, do: b
    neighbors |> Enum.filter(&(! MapSet.member?(visited, &1)))
      |> Enum.reduce(MapSet.put(visited, start), &(dfs(&2, adj, &1)))
  end
end

Main.main
